test.py

```diff
@@ -0,0 +1,108 @@
+"""
+Compare the output of the current Securify with past outputs.
+
+Author: Quentin Hibon
+
+Copyright 2018 ChainSecurity AG
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+"""
+import json
+import shutil
+import subprocess
+import tempfile
+
+from pathlib import Path
+
+import psutil
+
+
+class OutputMismatchException(Exception):
+    """Raised when there is a mismatch between past and current output
+    """
+    pass
+
+
+def test_securify_analysis(c_file, json_output, memory=4, overwrite=False):
+    """Compare the output of Securify on c_file with its expected output
+
+    Args:
+        c_file: the contract to analyze
+        json_output: the expected json_output
+        memory: the number of GB to use
+        overwrite: whether to overwrite the output in case of mismatch
+    """
+    mem_available = psutil.virtual_memory().available // 1024 ** 3
+    assert mem_available >= memory, f'Not enough memory to run: {mem_available}G'
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        output = Path(tmpdir) / 'sec_output.json'
+
+        cmd = ['java',
+               f'-Xmx{memory}G',
+               '-jar', 'build/libs/securify-0.1.jar',
+               '-fs', c_file,
+               '-o', output]
+
+        try:
+            print('Running:')
+            print(' '.join(str(o) for o in cmd))
+            subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        except subprocess.CalledProcessError as exn:
+            print('Securify already failed during execution!')
+            raise exn
+
+        if overwrite:
+            print('Overwriting.')
+            shutil.copy(output, json_output)
+            return
+
+        with open(output) as fsc:
+            current_output = json.load(fsc)
+
+        with open(json_output) as fsj:
+            expected_output = json.load(fsj)
+
+        if current_output != expected_output:
```

Comment:
Let's make this less stringent. Ideally I don't want all the tests to break if we fix stuff in Securify, e.g. like the securify errors currently inside the jsons, and adding new patterns.

-> Can we just compare if the new output still has the same results for existing patterns (both the nothing-found and the ones where stuff was found)? Basically a check if all elements inside `results` still exist inside the new results?

Alternatively we can also say that we expect all new patterns to return no issues with these files, and only break if we actually detect new issues.

Reply:
I was expecting that we would update the results in this case, so you'd prefer to check just a few fields?

Reply:
Actually I think it makes sense to compare the Securify errors, precisely to track the progress which is made, e.g., on this branch the empty contract fails to be analyzed, but #52  fixes this. Having it in the test allows us to check that the PR actually does something useful (on top of not introducing regressions). Also, if we don't compare errors, we could end up introducing new exceptions without knowing.

I still think choosing the fields we compare could make sense, especially given #59 which changes the output with text fields that are probably not worth testing anyway.

Reply:
Ok, comparing errors makes actually a lot of sense, you are right. So perhaps let's just compare the issue keys `violations`, `warnings`, safe`, `conflicts`, and implicitly assume that they must be empty if they don't exist at all for that pattern in the "correct" json output. This way we can add new patterns without having to update all the json with "empty" data, making the PRs where we add new patterns more readable as we will only add lines which actually mean something.